### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by TheSomeMan**

### CODE REVIEW

File: src/json_stream_gen.c [hunk 1]
```diff
@@ -0,0 +1,1478 @@
...
+    return *p_lc->decimal_point;
+}
+
+typedef struct json_stream_gen_float_str_buf_t
+{
+    char buffer[JSON_STREAM_GEN_STR_BUF_SIZE_FLOAT];
+} json_stream_gen_float_str_buf_t;
+
+static bool
+json_stream_gen_float_to_str(
```
Comment (ojousima) [inline src/json_stream_gen.c]: I would add the decimal separator as an argument to the function creating the string. 

As a use-case example elsewhere, in the Ruuvi Cloud we convert floats to strings by user preference. The precision and locale comes from user settings, not system locale. Passing the separator as an argument gives end-user more flexibility to decide if they want to use system locale, user setting, fixed "." or fixed ",".

Most importantly, this change means that the output will not depend on system running the code unless user explicitly wants it to depend on locale.

File: src/json_stream_gen.c [hunk 2]
```diff
@@ -0,0 +1,1478 @@
...
+                float_str.buffer))
+        {
+            return false;
+        }
+    }
+    p_gen->is_first_item = false;
+    return true;
+}
+
+typedef struct json_stream_gen_double_str_buf_t
```
Comment (ojousima) [inline src/json_stream_gen.c]: Same comment about locale

**Review by ojousima — APPROVED** [review]
Thanks, the project is really well documented and handles a wide variety of use cases. 

Please see the comment about number localization, final decision on if you want to make that change is yours.

File: src/json_stream_gen.c (hunk 1 above)
Comment (TheSomeMan) [reply src/json_stream_gen.c]: I checked the standards related to JSON format and they state that only '.' can be used as a decimal separator (so, comma is not allowed).
image

The following code is used to identify the character used as a decimal separator in the printf function. Once the float or double value is printed, this decimal separator is replaced by a period ('.').
```
static char
json_stream_gen_get_decimal_point(void)
{
    struct lconv* const p_lc = localeconv();
    if (NULL == p_lc)
    {
        return '.';
    }
    return *p_lc->decimal_point;
}

...

{
    // ...
    
     len = snprintf(p_str->buffer, sizeof(p_str->buffer), "%1.*g", precision, val);

    // ...
    
    char* p_decimal_point = strchr(p_str->buffer, json_stream_gen_get_decimal_point());
    if (NULL != p_decimal_point)
    {
        *p_decimal_point = '.';
    }
}
```

Comment (ojousima) [reply src/json_stream_gen.c]: Thanks, then I misunderstood the function. I thought it was printing float as a string customized to user locale, such as 5.324 -> "5,32"

**Pull request merged**